script/engine.go

```diff
@@ -746,6 +978,68 @@ func (e *Engine) runCommand(s *State, cmd *command, impl Cmd) error {
 		s.Flags = fs
 	}
 
+	if fs.Changed("autocomplete") {
+		if usage.AutocompleteArgs == nil && usage.AutocompleteFlag == nil {
+			return nil
+		}
+
+		fs := pflag.NewFlagSet(cmd.name, pflag.ContinueOnError)
+		if usage.Flags != nil {
+			usage.Flags(fs)
+		}
+
+		argOrFlags := autocompleteParse(*autocomplete, fs)
+		if len(argOrFlags) == 0 {
+			return nil
+		}
+
+		var before []string
+		for i := 0; i < len(argOrFlags)-1; i++ {
+			if argOrFlags[i].arg != "" {
+				before = append(before, argOrFlags[i].arg)
+			}
+		}
+
+		var suggestions []string
+		last := argOrFlags[len(argOrFlags)-1]
+		if last.argOrFlag == 1 {
+			if usage.AutocompleteFlag == nil {
+				return nil
+			}
+
+			// Cursor is in a flag value
+			if last.hasValue {
+				for _, suggestion := range usage.AutocompleteFlag(s, before, last.flagName, last.flagValue) {
+					suggestions = append(suggestions, cmd.name+" "+(*autocomplete)[:len(*autocomplete)-len(last.flagValue)]+suggestion)
+				}
+			} else {
+				// If the current input is not a valid flag name
+				if fs.Lookup(last.flagName) == nil {
+					fs.VisitAll(func(f *pflag.Flag) {
+						if last.flagName == "" || strings.HasPrefix(f.Name, last.flagName) {
+							suggestions = append(suggestions, cmd.name+" "+(*autocomplete)[:len(*autocomplete)-len(last.flagName)]+f.Name)
+						}
+					})
+				}
+			}
+		} else {
+			if usage.AutocompleteArgs == nil {
+				return nil
+			}
+
+			for _, suggestion := range usage.AutocompleteArgs(s, before, last.arg) {
+				suggestions = append(suggestions, cmd.name+" "+(*autocomplete)[:len(*autocomplete)-len(last.arg)]+suggestion)
+			}
+		}
+
+		s.stdout = strings.Join(suggestions, "\n") + "\n"
+		if s.stdout != "" {
```

Comment:
this condition is always true (there's always at least `\n`)? also why put these into the stdout buffer and write them with the `[stdout]` header?

Reply:
Ah, yes, we can remove the conditional. I place it in the buffer and log because that is also what we do for for normal command output. I figured there was a reason, but I don't know what it is.

```
        stdout, stderr, waitErr := wait(s)
	s.stdout = stdout
	s.stderr = stderr
	if stdout != "" {
		s.Logf("[stdout]\n%s", stdout)
	}
	if stderr != "" {
		s.Logf("[stderr]\n%s", stderr)
	}
```